build-system/babel-plugins/babel-plugin-transform-log-methods/index.js

```diff
@@ -0,0 +1,320 @@
+ */
+const defaultMessagesPath = messagesPath;
+
+/** Method on `Log` that this transform outputs to "expand" log messages. */
+const messageExpansionMethod = 'expandLogMessage';
+
+/**
+ * Mangled expansion method name to prevent large public method name in
+ * multipass build.
+ */
+const messageExpansionMethodShortName = 'l';
+
+/**
+ * Approximate length of a compressed message expansion call to determine
+ * whether an instance of message indirection actually reduces binary size.
+ */
+const roughMinifiedExpansionLength = 'a().b("xx")'.length;
+
+/**
+ * Determines a filepath relative to the repo root.
+ * @param {string} path
+ * @return {string}
+ */
+const relativeToRoot = path => `${__dirname}/../../../${path}`;
+
+/**
+ * Reads the messages table from the JSON file.
+ * @param {string} messagesPath
+ * @return {!Object<string, string>}
+ */
+function getMessages(messagesPath) {
+  try {
+    return JSON.parse(fs.readFileSync(messagesPath));
+  } catch {
+    // When non-existent or empty just return an empty object. Transformations
+    // that read will write as well.
+    return {};
+  }
+}
+
+/**
+ * Writes the messages table to the JSON file.
+ * @param {string} messagesPath
+ * @param {!Object<string, string>} obj
+ */
+function writeMessages(messagesPath, obj) {
+  const json = JSON.stringify(obj, /* replacer */ null, /* spaces */ 2);
+  const messagesPathDir = path.dirname(messagesPath);
+  if (!fs.existsSync(messagesPathDir)) {
+    fs.mkdirSync(messagesPathDir);
+  }
+  fs.writeFileSync(messagesPath, json);
+}
+
+/**
+ * Gets a message id from the table, or adds a new entry for a message if
+ * non-existent and returns its new id.
+ * @param {string} messagesPath
+ * @param {string} message
+ * @return {string} Short message id.
+ */
+function getOrCreateShortMessageId(messagesPath, message) {
+  const messages = getMessages(messagesPath);
+  if (messages[message]) {
+    return messages[message];
+  }
+  // Base-36 radix for best utilization of ascii space.
+  const shortMessageId = Object.keys(messages).length.toString(36);
+  messages[message] = shortMessageId;
+  writeMessages(messagesPath, messages);
+  return shortMessageId;
+}
+
+/**
+ * Builds a message template using printf syntax from a starting node.
+ * @param {@babel/types} t
+ * @param {!Node} node
+ * @param {!Array<!Node>} interpolationArgs
+ * @return {string} Template for the message.
+ */
+function buildMessage(t, node, interpolationArgs) {
+  if (t.isStringLiteral(node)) {
+    return node.value;
+  }
+
+  if (t.isBinaryExpression(node, {operator: '+'})) {
+    return (
+      buildMessage(t, node.left, interpolationArgs) +
+      buildMessage(t, node.right, interpolationArgs)
+    );
+  }
+
+  if (t.isTemplateLiteral(node)) {
+    let quasied = '';
+    let i = 0;
+    for (; i < node.quasis.length - 1; i++) {
+      quasied += node.quasis[i].value.cooked;
+      quasied += buildMessage(t, node.expressions[i], interpolationArgs);
```

Comment:
Why are we building with the dynamic expression, too? Are people actually passing in string literals (eg, ``` `before${'foo'}after` ```)?

Reply:
Not sure. @erwinmombay wrote this originally. I'd say there's no harm in doing this recursively?

Reply:
I think the concerns are mixed. It should just be a `%s` that we put in the message, bu it's also putting the args into the `interpolationArgs`. Oh well.